### PR TITLE
78 - Chapter and Nonprofit Deletion

### LINKED ISSUE #78: [Backend] Ability to remove chapters, non-profits
## What/Why?
Admin-only feature to remove chapters, non-profits from the system.

#### Chapter
1. API call must completely delete any information pertaining to the chapter.
2. If not already existing, create a table which stores name of chapter that is removed, and timestamp of removal - and populate table on each removal.
3. Email chapter notifying they have been removed due to X reason (String input).

#### Non-Profit
1. API call must completely delete any information pertaining to the non-profit.
2. If not already existing, create a table which stores name of non-profit that is removed, and timestamp of removal - and populate table on each removal.
3. Email non-profit notifying they have been removed due to X reason (String input).

### PR DESCRIPTION
**Pull request by Navbryce**
## What is this?

Delete chapters and nonprofits; cascade deletion to dependencies. Creates a log of all chapter and nonprofit deletions. Sends email upon deletion.

## Steps to Test

- Tested locally
## Associated Issues

Closes #78

### CODE REVIEW

File: server/mongodb/models/Chapter.ts [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
+import UserModel from "server/mongodb/models/User";
+import { deleteDocumentAndDependencies } from "server/utils/delete-document";
 import { Chapter, MaintenanceType } from "src/utils/types";
 
 const ChapterSchema = new Schema<Chapter>({
   name: {
     type: String,
     required: true,
   },
-  email: String,
```
Comment (Navbryce) [inline server/mongodb/models/Chapter.ts:10]: This probably should have been deleted when the `contact` field was added

File: server/utils/delete-document.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * @param model - the model of the document
+ * @param filter - the filter that selects the documents to be deleted
+ *
+ * @return - the number deleted
+ */
+export async function deleteDocumentAndDependencies<T>(
+  model: Model<T> & { hasNoDependencies?: () => boolean },
+  filter: FilterQuery<T>
+): Promise<number> {
+  if (model.hasNoDependencies && model.hasNoDependencies()) {
```
Comment (Navbryce) [inline server/utils/delete-document.ts:21]: This is an optimization if deletes don't need to be cascaded

Comment (Navbryce) [inline server/utils/delete-document.ts:21]: Optimization if deletes don't need to be cascaded

File: src/utils/types.ts [hunk 3]
```diff
@@ -300,7 +310,12 @@ export interface UploadedFile {
...
 }
 
+// BaseEmailTemplateData holds the required values for the base email
+export interface BaseEmailTemplateData {
+  name: string;
+}
+
 export interface Email {
-  data: Record<string, unknown>; // data contains the the variables for the template
+  data: Record<string, unknown> & BaseEmailTemplateData; // data contains the the variables for the template
```
Comment (Navbryce) [inline src/utils/types.ts:319]: Assuming every Email extends the BaseTemplate, they all need a "name" value supplied.

File: server/mongodb/models/Chapter.ts (hunk 1 above)
Comment (mjbernet3) [reply server/mongodb/models/Chapter.ts:10]: Yeah I think I agree with that.  Thoughts @shashanklal01?

Comment (shashanklal01) [reply server/mongodb/models/Chapter.ts:10]: Sounds good to me

File: server/mongodb/models/Chapter.ts [hunk 4]
```diff
@@ -31,6 +33,23 @@ const ChapterSchema = new Schema<Chapter>({
...
+ChapterSchema.pre(
+  "remove",
+  async function (next: (err?: CallbackError) => void) {
+    const id = this._id as Types.ObjectId;
+    // Cascade deletes (and remove chapter from projects)
+    await Promise.all([
+      // delete the user associated with the nonprofit
+      deleteDocumentAndDependencies(UserModel, {
+        chapter: id,
+      }),
```
Comment (shashanklal01) [inline server/mongodb/models/Chapter.ts:45]: are you deleting the user that creates the nonprofit page, or the nonprofit alone?

File: server/mongodb/models/Nonprofit.ts [hunk 5]
```diff
@@ -24,6 +27,25 @@ const NonprofitSchema = new Schema<Nonprofit>({
...
+NonprofitSchema.pre(
+  "remove",
+  async function (next: (err?: CallbackError) => void) {
+    const id = this._id as Types.ObjectId;
+    // Cascade deletes
+    await Promise.all([
+      // delete the user associated with the nonprofit
+      deleteDocumentAndDependencies(UserModel, {
+        nonprofit: id,
+      }),
```
Comment (shashanklal01) [inline server/mongodb/models/Nonprofit.ts:39]: same question as before

File: src/actions/Chapter.ts [hunk 6]
```diff
@@ -16,3 +16,13 @@ export async function chapterUpdateChapter(
     },
   });
 }
+
+export async function adminDeleteChapter(chapterId: string, reason: string) {
+  return internalRequest({
+    url: chapterAPI + `/${chapterId}/chapter`,
```
Comment (shashanklal01) [inline src/actions/Chapter.ts:22]: I think its more intuitive to keep the URLs for admin functionality to be /api/natlAdmin/...

File: src/actions/Nonprofit.ts [hunk 7]
```diff
@@ -33,3 +33,16 @@ export async function nonprofitUpdateNonprofit(
     },
   });
 }
+
+export async function adminDeleteNonProfit(
+  nonprofitId: string,
+  reason: string
+) {
+  return internalRequest({
+    url: nonprofitAPI + `${nonprofitId}/nonprofit`,
```
Comment (shashanklal01) [inline src/actions/Nonprofit.ts:42]: same as above

**Comment by shashanklal01** [comment]
I'm going to merge the other backend ticket first, so I'll hold off of this as I expect merge conflicts

File: server/mongodb/models/Chapter.ts (hunk 4 above)
Comment (Navbryce) [reply server/mongodb/models/Chapter.ts:45]: This deletes all users associated with the chapter (e.g., have the chapter field populated with this id)

File: server/mongodb/models/Nonprofit.ts (hunk 5 above)
Comment (Navbryce) [reply server/mongodb/models/Nonprofit.ts:39]: This deletes all users associated with the nonprofit (e.g., the nonprofit field has this id).

File: emails/templates/chapter-removal/html.pug [hunk 8]
```diff
@@ -0,0 +1,4 @@
+extends ../../shared/base-email.pug
+block content
+    p The chapter you represent, #{chapterName}, has been removed from the National Hack4Impact Portal for the following reason:
+    p #{reason}
```
[the remark this answers is not in the document]
Comment (Navbryce) [reply emails/templates/chapter-removal/html.pug:4]: Sounds good. I'll just add a generic reason into the frontend actions, so we don't need to re-do this work on the backend when we add that in.

**Pull request merged**